File: blackwidow/engine/widgets/image_widget.py
from django.forms import ClearableFileInput, CheckboxInput
from django.utils.html import conditional_escape
from django.utils.safestring import mark_safe

__author__ = 'Ziaul Haque'


class ImageWidget(ClearableFileInput):
    """
        renders the input file as an thumbnail image, and modify the 'default html'
    """
    template_with_initial = u'%(initial)s %(clear_template)s<br />%(input)s'
    template_with_clear = '<div style="display:inline">%(clear)s ' \
                          '<label for="%(clear_checkbox_id)s">%(clear_checkbox_label)s</label></div>'

    def render(self, name, value, attrs=None):
        substitutions = {
            'input_text': self.input_text,
            'clear_template': '',
            'clear_checkbox_label': self.clear_checkbox_label,
        }
        template = u'%(input)s'
        _input = super(ClearableFileInput, self).render(name, value, attrs)
        _input_div = '<div class="custom-upload">' + _input +\
                     '<div class="fake-file"><input disabled="disabled" style="margin: 0 0 0 160px"></div></div>'
        substitutions['input'] = _input_div

        template = self.template_with_initial
        _url = '/static/img/dummy-photo.png'
        dummy_img_tag = '<img class="thumbnail-border" height="100" width="100" src="' + _url + '" />'
        substitutions['initial'] = (u'%s' % dummy_img_tag)

        if self.is_initial(value):
            substitutions['initial'] = (u'%s' % (value.instance.__str__()))
            if not self.is_required:
                checkbox_name = self.clear_checkbox_name(name)
                checkbox_id = self.clear_checkbox_id(checkbox_name)
                substitutions['clear_checkbox_name'] = conditional_escape(checkbox_name)
                substitutions['clear_checkbox_id'] = conditional_escape(checkbox_id)
                substitutions['clear'] = CheckboxInput().render(checkbox_name, False,
                                                                attrs={'id': checkbox_id, 'class': 'clear_image'})
                substitutions['clear_template'] = self.template_with_clear % substitutions

        return mark_safe(template % substitutions)

    def is_initial(self, value):
        return bool(value and hasattr(value, 'instance'))

File: static/js/widget/image_widget.js
/**
 * Created by Ziaul Haque on 8/28/2016.
 */


$(document).ready(function () {
    function _previewImage(input) {
        if (input.files && input.files[0]) {
            var reader = new FileReader();
            reader.onload = function (e) {
                $(input).parent().parent().find('img').attr('src', e.target.result);
            };
            reader.readAsDataURL(input.files[0]);
        }
    }

    var $form = $('form');

    $form.on("change", "input[type='file']", function () {
        _previewImage(this);
    });

    function _clearImage(input) {
        // '<img class="thumbnail-border" height="100" width="100" src="/static/img/dummy-photo.png" />'
        if ($(input).is(':checked') == true) {
            $(input).data('src', $(input).parent().parent().find('img').attr('src'));
            $(input).parent().parent().find('input:file').attr('disabled', 'disabled');
            $(input).parent().parent().find('img').attr('src', '/static/img/dummy-photo.png');
        } else {
            $(input).parent().parent().find('input:file').removeAttr('disabled');
            $(input).parent().parent().find('img').attr('src', $(input).data('src'));
        }

    }

    $form.on("change", "input[class='clear_image']", function () {
        _clearImage(this);
    });

    $('.fake-file').find('input').val('No file chosen');
    $form.on("change", ".custom-upload input[type=file]", function () {
        var file_path = $(this).val();
        if (file_path == undefined || file_path == "") {
            $(this).parent().find('.fake-file input').val("No file chosen");
        }
        else {
            var _file = file_path.substring(file_path.lastIndexOf("\\") + 1);
            $(this).parent().find('.fake-file input').val(_file);
        }
    });
});

